{% extends "base.html" %}

{% set active_page = "help" %}

{% block content %}
<div class="container">
    <h2>What is MarkerMiner?</h2>
    <p class="lead">MarkerMiner is a pipeline which identifies a set of single-copy (SC) genes in the input assembled transcriptomic datasets (for example: 1KP data) by comparing against a database of known SC genes, compiled from a systematic survey of duplication-resistant genes in the sequenced genomes of 20 flowering plants (<a href="http://www.pnas.org/content/110/8/2898.full" target="_blank">De Smet et al. 2013, PNAS&nbsp;<span class="glyphicon glyphicon-new-window"></span></a>)</p>
    <p>The reference database of single-copy genes is composed of Orthologous Groups (OGs), built using the OrthoMCL method. OGs are classified as being single-copy if they are present in all angiosperm species, tolerating any missing copies in up to two species and allowing for duplicates in up to three species. These adjustments were made to accommodate for the presence of recent duplicates, alleles, or pseudogenes and also for variations in the reference genome annotation. They are flagged as "<samp>Strictly</samp>" single-copy if OGs are truly SC for all species or "<samp>Mostly</samp>" single-copy with with duplicates for one up to three species.</p>
    <p>Clusters of single-copy taranscripts present in the input transcriptomic assemblies are identified by aligning and filtering against a chosen reference proteome database, which generates a tabular report of identified SC transcripts. These clusters are run through a multiple sequence alignment (MSA) step to generate an alignment profile, which can help assess sequence variation and conduct preliminary phylogenetic analysis. The MSA profile is mapped to the spliced coding sequence of the reference gene to generate a spliced aligned, which can prove to be useful for primer/probe design.</p>
    <h2>How does the pipeline work?</h2>
    <p>The pipeline, takes as input, single (or multiple) transcriptome assemblies and processes them through the pipeline, which can be separated into 3 major steps:</p>
    <ol>
        <li>Following the length filtering of the transcripts (<kbd>minTranscriptLen &gt;= 900</kbd>), BLASTX of the transcripts against the complete proteome of the chosen reference organism and the TBLASTN of the reference proteome against the transcript assemblies is conducted. The results of the BLASTX and TBLASTN are processed to only retain top hits with a <kbd>minTranscriptCoverage &gt;= 70%</kbd> and a <kbd>minProteinCoverage &gt;= 80%</kbd>, respectively while maintaining a <kbd>minSimilarity &gt;= 80%</kbd>. Based on this bi-directional BLAST analysis, only the reciprocally unique best hits are retained. Finally, the BLAST hits are filtered to only retain the aligning transcripts, matched against the list of known single-copy genes in the reference proteome.</li>
        <li>Using the above filtered BLAST results, the putative single-copy (SC) transcripts are clustered with the matching single-copy gene from the reference proteome. The transcripts are then re-oriented according to the strandedness of the gene from the reference proteome. The results of this clustering are recorded in an output file (<code>single_copy_genes.out</code>), consisting of the following set of columns:</li>
            <ol>
                <li><samp>'Identifier'</samp> of reference proteome gene</li>
                <li><samp>'Single copy status'</samp> (i.e. Mostly/Strictly)</li>
                <li><samp>'Gene Description'</samp> of reference proteome gene</li>
                <li><samp>'Number of orthologs (N)'</samp> across the input transcriptomic assemblies</li>
                <li><samp>'N'</samp> columns for each input transcriptome with a value of either <samp>'OrganismCode_ScaffoldId_OrganismName_Orientation</samp>' or <samp>'NA</samp>' (when no ortholog was identified)</li>
            </ol>
        <li>The above clustered re-oriented transcript sequences are run through a <samp>mafft</samp> multiple sequence alignment (MSA) step to generate an MSA profile alignment. This profile serves as an input to the <samp>muscle</samp> tool which maps this profile to a spliced coding sequence of the reference gene (with the introns replaced by a stretch of <samp>NNNNs</samp>). The output is a refined MSA profile alignment with gaps in place of putative introns.</li>
    </ol>
</div>
{% endblock %}
